Reset directory to /docker-entrypoint-initdb.d for each initdb script

JacobHayes opened the pull request:
Since we source the scripts in initdb.d, any `cd`s will affect other scripts. They could explicitely `cd` back to `/docker-entrypoint-initdb.d`, but I think most would expect to already be there.

Alternatively, if there's not a use case for sourcing the scripts, we could run scripts instead. Not sure which (or either) is more fitting behavior.

JacobHayes commented:
Those failures seem to be related to a short amount of github downtime; I noticed it timing out in chrome a bit ago. Any way to restart them?

JacobHayes commented:
Actually, the more I think about it (and write initdb scripts), I think sourcing is the wrong pattern. Sourcing will leak script functions and actions into each other (and main script). Maybe worse yet, they will all inherit any `set` commands, including the `set -e` in the main script, possibly into scripts not written with that in mind. Additionally, because we conditionally source the files, it affects using `set -e` in the sourced script as described [here](http://stackoverflow.com/a/4073372)

Thoughts?

tianon commented:
Interesting conundrum -- honestly, I see the ability of these scripts to interact as one of their stronger features (and the way they automatically inherit the `WORKDIR` set in the `Dockerfile` or at runtime is sometimes useful).  One snippet could be responsible for defining a load of useful functions and variables, while later scripts refer back to those, so I'm actually thinking something more like this might be interesting, to at least insulate our own script's environment from changes made by these third-party scripts:

```diff
diff --git a/docker-entrypoint.sh b/docker-entrypoint.sh
index 3b436e4..b2148c8 100755
--- a/docker-entrypoint.sh
+++ b/docker-entrypoint.sh
@@ -75,16 +75,18 @@ if [ "$1" = 'postgres' ]; then
 
 		psql+=( --username "$POSTGRES_USER" --dbname "$POSTGRES_DB" )
 
-		echo
-		for f in /docker-entrypoint-initdb.d/*; do
-			case "$f" in
-				*.sh)     echo "$0: running $f"; . "$f" ;;
-				*.sql)    echo "$0: running $f"; "${psql[@]}" < "$f"; echo ;;
-				*.sql.gz) echo "$0: running $f"; gunzip -c "$f" | "${psql[@]}"; echo ;;
-				*)        echo "$0: ignoring $f" ;;
-			esac
+		(
 			echo
-		done
+			for f in /docker-entrypoint-initdb.d/*; do
+				case "$f" in
+					*.sh)     echo "$0: running $f"; . "$f" ;;
+					*.sql)    echo "$0: running $f"; "${psql[@]}" < "$f"; echo ;;
+					*.sql.gz) echo "$0: running $f"; gunzip -c "$f" | "${psql[@]}"; echo ;;
+					*)        echo "$0: ignoring $f" ;;
+				esac
+				echo
+			done
+		)
 
 		gosu postgres pg_ctl -D "$PGDATA" -m fast -w stop
 
```

This still preserves things like the fact that our `psql` array is available for use within these scripts (`"${psql[@]}" ...`), but ensures that we don't inadvertently lose our own `set -e` or gain a `set -x` due to one of these scripts (except for the duration of the loop there, which is fairly unavoidable).

Having it source instead of run also means that if someone wants to use a `*.sql` file to seed their database but they aren't concerned about failure that they can add a simple script that runs before their `*.sql` file is loaded that does a `set +e` to disable what we've set, ensuring that the failure is ignored without writing the simple import code themselves.

Now, this _does_ prevent these scripts from doing anything to `PGDATA`, but I admit I don't fully understand the use case for modifying `PGDATA` in one of these scripts.  Can you elaborate?

mmarzantowicz commented:
Wouldn't it be better to `cd` in line 81 (https://github.com/JacobHayes/postgres/blob/reset-after-initdb-scripts/9.1/docker-entrypoint.sh#L81) just after the entry to the for loop?

Why can't you `cd` inside your script in `/docker-entrypoint-initdb.d/`? I prefer to work using absolute paths (I think there are some other people like me) and using `cd` in any scripts is no go for me.

JacobHayes commented:
I imagined shared resources would be sourced manually. They couldn't be in the call path (omit extension or outside initdb.d or side effect free), but could still be easily referenced. Wouldn't be able to modify values between scripts easily that way though. Regarding inheriting WORKDIR, that doesn't seem to happen now:
``` shell
/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/1_pwd.sh
/docker-entrypoint-initdb.d
```

@tianon I like the `()` for insulating the main script. Thoughts on the above and need for sourcing in the entrypoint? If you're still in favor, I'll add what you have and call it good. I was modifying `PGDATA` to move config files outside of the data directory. Since PGDATA needs `postgresql.conf` (which can then specify an actual `data_directory`), I moved the config files and changed PGDATA (after initdb ran) to point to the config directory. I think instead I should set PGDATA in my Dockerfile, and then move things to and set `data_directory`. As an aside, it doesn't seem that pg_ctl follows `data_directory`, so also need to a link to `postmaster.pid` (and maybe `postmaster.opts`) in the config directory.

@mmarzantowicz Yes, that would be better, don't know my reasoning otherwise. :) Yes I can, but I had expected to be in `/docker-entrypoint-initdb.d/` (as is the case for only one script) no matter what order my scripts ran, and figured I wasn't alone in that. Preference for absolute paths doesn't preclude others from following that though, as was done in the [postgis image](https://github.com/appropriate/docker-postgis/blob/f7057086d58a40ba625f7d445c986029090e5f0b/initdb-postgis.sh#L15) before (tangentially) removing 9.0 supporting code. Easier to solve it here than every downstream script that uses `cd` or has to work around it.
